torch/csrc/autograd/function.h

```diff
@@ -257,11 +283,62 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   // Miscellaneous Methods
   //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  /// The sequence number of this `Node`.
+  /// NOTE [ Sequence Number]
+  ///
+  /// The sequence_nr has two main usages in autograd:
+  ///
+  /// 1) Determine the node's execution priority in the engine.
+  ///    Nodes with higher priority numbers are executed first. There is actually no
+  ///    particular reason why we want nodes corresponding to ops executed later to be
+  ///    the first to be executed in the backward pass. The point is to simply make
+  ///    it *deterministic* so its a bit easier to reason about.
+  ///    One caveat is that sequence_nr is UINT64_MAX for AccumulateGrad nodes
+  ///    because we want to prioritize them (why?)
```

Comment:
Because they are always leafs, there is no reason to delay them.
Also in the case where the `.grad` field already exists, just before the AccumulateGrad Node is a bad time as you have both the current grad buffer in the engine and the one in the .grad field.
So you want to execute the AccumulateGrad Node asap to be able to accumulate into the .grad field and drop the extra buffer that the engine is holding onto.

Reply:
Hmm that makes sense - my question is now, when I execute the typical non-accumulate grad Node, does my memory usage decrease/increase stay the same? For one, if there are saved variables, they would get released in the backward pass, but there is also the chance that if you have multiple children, you'd be allocating more buffers. 

So it seems like executing nodes with one child are guaranteed to decrease memory utilization, but executing nodes with multiple children may keep memory usage the same or possibly increase.

Reply:
For other Nodes we cannot predict that. Mostly because input size(s) are not linked to output size(s). Like `.sum()` will be very bad as the grad_input will be huge compared to the output. While expand will be nice because it will be smaller. But the engine has no way to know the difference between the two.

Another argument for the AccumulateGrad is for DDP.
Basically they start syncing the gradients across machines as soon as they are ready (the AccumulateGrad Node is called). So the earlier we call it, the more time we have to overlap transfer of grads with the remainder of the backward pass.